shared/versions.yml

```diff
@@ -7,126 +7,84 @@
     tag: '22'
 couchdb_prometheus_exporter_p2sd:
   upstream_image: gcr.io/google-containers/prometheus-to-sd:v0.3.2
-  image: gcr.io/gpii2test-common-stg/gcr.io/google-containers/prometheus-to-sd
-  sha: sha256:7f7e424211b64a77b5ee2e8a3085048b54c1676217ee8830dc4112df6111b8a2
-  tag: v0.3.2
   generated:
     image: gcr.io/gpii2test-common-stg/gcr.io/google-containers/prometheus-to-sd
     sha: sha256:7f7e424211b64a77b5ee2e8a3085048b54c1676217ee8830dc4112df6111b8a2
     tag: v0.3.2
 couchdb:
   upstream_image: couchdb:2.2.0
-  image: gcr.io/gpii2test-common-stg/couchdb
-  sha: sha256:7d60986a0b43015711121909c50b94dd6a1a6520f8f4bc72f52109b9ea8a17ad
-  tag: 2.2.0
   generated:
     image: gcr.io/gpii2test-common-stg/couchdb
     sha: sha256:7d60986a0b43015711121909c50b94dd6a1a6520f8f4bc72f52109b9ea8a17ad
     tag: 2.2.0
 dataloader:
   upstream_image: gpii/universal:latest
-  sha: sha256:eab9307f4b1549077c97646848e54b49fc6ceed42d98185882e2258b409f0770
-  image: gcr.io/gpii2test-common-stg/gpii/universal
-  tag: latest
   generated:
     image: gcr.io/gpii2test-common-stg/gpii/universal
     sha: sha256:eab9307f4b1549077c97646848e54b49fc6ceed42d98185882e2258b409f0770
     tag: latest
 flowmanager:
   upstream_image: gpii/universal:latest
-  sha: sha256:eab9307f4b1549077c97646848e54b49fc6ceed42d98185882e2258b409f0770
-  image: gcr.io/gpii2test-common-stg/gpii/universal
-  tag: latest
   generated:
     image: gcr.io/gpii2test-common-stg/gpii/universal
     sha: sha256:eab9307f4b1549077c97646848e54b49fc6ceed42d98185882e2258b409f0770
     tag: latest
 k8s_snapshots:
   upstream_image: elsdoerfer/k8s-snapshots:v2.0
-  image: gcr.io/gpii2test-common-stg/elsdoerfer/k8s-snapshots
-  sha: sha256:8723828551f49970389ed024e3073fd1c5425d1bd2eacb286925a67d381ec828
-  tag: v2.0
   generated:
     image: gcr.io/gpii2test-common-stg/elsdoerfer/k8s-snapshots
     sha: sha256:8723828551f49970389ed024e3073fd1c5425d1bd2eacb286925a67d381ec828
     tag: v2.0
 locust:
   upstream_image: gpii/locust:0.9.0-gpii.2
-  image: gcr.io/gpii2test-common-stg/gpii/locust
-  sha: sha256:e526da4f2243d7d0a330230f3c94be04ed558d0ed8946484faf7fce02ba95657
-  tag: 0.9.0-gpii.2
   generated:
     image: gcr.io/gpii2test-common-stg/gpii/locust
     sha: sha256:e526da4f2243d7d0a330230f3c94be04ed558d0ed8946484faf7fce02ba95657
     tag: 0.9.0-gpii.2
 nginx_ingress:
   upstream_image: quay.io/kubernetes-ingress-controller/nginx-ingress-controller:0.15.0
-  image: gcr.io/gpii2test-common-stg/quay.io/kubernetes-ingress-controller/nginx-ingress-controller
-  sha: sha256:0489cd1c6f0755f532d4b8204d5dabea038d4b2be8ec6efc2da584ff3f5aba61
-  tag: 0.15.0
   generated:
     image: gcr.io/gpii2test-common-stg/quay.io/kubernetes-ingress-controller/nginx-ingress-controller
-    sha: sha256:7b79e1bc6437e6376dadf558e012adde6395bb28dee4a38ce08c7e5c9f220178
+    sha: sha256:0489cd1c6f0755f532d4b8204d5dabea038d4b2be8ec6efc2da584ff3f5aba61
```

Comment:
Hmm, I'll have to look at why this sha is changing. It shouldn't be, according to https://quay.io/repository/kubernetes-ingress-controller/nginx-ingress-controller?tag=latest&tab=tags.

Reply:
Weird, it seems that sometimes the sha256 is different for the re-pushed image. But only sometimes! Here's a recent result:

```
/home/app/gpii-infra # docker images | grep gcr.io | awk '{print $3}' | xargs docker inspect | jq .[].RepoDigests
[
  "busybox@sha256:954e1f01e80ce09d0887ff6ea10b13a812cb01932a0781d6b0cc23f743a874fd",
  "gcr.io/gpii2test-common-stg/busybox@sha256:f79f7a10302c402c052973e3fa42be0344ae6453245669783a9e16da3d56d5b4"
]
[
  "gpii/universal@sha256:eab9307f4b1549077c97646848e54b49fc6ceed42d98185882e2258b409f0770",
  "gcr.io/gpii2test-common-stg/gpii/universal@sha256:eab9307f4b1549077c97646848e54b49fc6ceed42d98185882e2258b409f0770"
]

...
...

[
  "kocolosk/couchdb-statefulset-assembler@sha256:a689a5cde2c19b7767c79fc2d265844bbc5913f10f5516480d7e06e2292f234e",
  "gcr.io/gpii2test-common-stg/kocolosk/couchdb-statefulset-assembler@sha256:a689a5cde2c19b7767c79fc2d265844bbc5913f10f5516480d7e06e2292f234e"
]
[
  "elsdoerfer/k8s-snapshots@sha256:8723828551f49970389ed024e3073fd1c5425d1bd2eacb286925a67d381ec828",
  "gcr.io/gpii2test-common-stg/elsdoerfer/k8s-snapshots@sha256:c787dc6fa1812c97ac96f46b4ff6dc1485e06c458839d4aa129aaf62201309e9"
]
```

I had implemented a "fix" for this previously (look for the expected gcr.io image name in RepoDigests instead of always taking the first one), but it looks like I never committed it (I didn't think I needed it due to the transient nature of this issue). I will re-do this fix, falling back to taking the first RepoDigest (I think this was missing from my first pass and lead me to discard the strategy).

Reply:
I think this is due to mixing manifest digests and image digests: `.RepoDigests` shows you a digest of the whole docker manifest, and this can be different locally and in the repo for several reasons (for example due to how layers are stored), that's why the `RepoDigests` is there. What should always stay the same is the Image (`Config.Image`) digest.

Reply:
Nice article about this is https://windsock.io/explaining-docker-image-ids/

Reply:
> What should always stay the same is the Image (Config.Image) digest.

Yes. Would be nice if we could use this value, but we can't:

```
/home/app/gpii-infra # docker inspect busybox | less | jq .[].Config.Image
"sha256:90b7037cc5e65fa9e3f33e8096febd6fad8af0ff94876d73dabe048d65bec645"

/home/app/gpii-infra # docker run -it busybox@sha256:90b7037cc5e65fa9e3f33e8096febd6fad8af0ff94876d73dabe048d65bec645 sh
Unable to find image 'busybox@sha256:90b7037cc5e65fa9e3f33e8096febd6fad8af0ff94876d73dabe048d65bec645' locally
INFO[2019-04-15T14:16:46.115518200Z] Attempting next endpoint for pull after error: manifest unknown: manifest unknown
docker: Error response from daemon: manifest for busybox@sha256:90b7037cc5e65fa9e3f33e8096febd6fad8af0ff94876d73dabe048d65bec645 not found.
See 'docker run --help'.
```